internal/dataplane/kong_client.go

```diff
@@ -439,6 +452,50 @@ func (c *KongClient) Update(ctx context.Context) error {
 	return nil
 }
 
+func (c *KongClient) FetchLastGoodConfiguration(ctx context.Context) error {
+	if c.lastValidKongState == nil {
+		gatewayClients := c.clientsProvider.GatewayClients()
+		c.logger.Debugf("sending configuration to %d gateway clients", len(gatewayClients))
+
+		type kongStateWithHash struct {
+			kongState *kongstate.KongState
+			hash      string
+		}
+
+		var goodKongState *kongStateWithHash
+		_, err := iter.MapErr(gatewayClients, func(client **adminapi.Client) (*kongStateWithHash, error) {
+			rs, err := dump.Get(ctx, (*client).AdminAPIClient(), dump.Config{})
+			if err != nil {
+				return nil, err
+			}
+			ks := kongstate.KongRawStateToKongState(rs)
+			status, err := c.configGetter.GetCurrentStatus(ctx, (*client).AdminAPIClient())
+			if err != nil {
+				return nil, err
+			}
+			(*client).SetLastConfigSHA([]byte(status.ConfigurationHash))
+			kongStateWithHash := &kongStateWithHash{
+				kongState: ks,
+				hash:      status.ConfigurationHash,
+			}
+			if status.ConfigurationHash != sendconfig.WellKnownInitialHash {
+				// get the first good one as the one to be used
+				goodKongState = kongStateWithHash
```

Comment:
Please note `iter.MapErr` runs the iterations concurrently and if I'm not mistaken we have a possibility of race condition here (concurrent writes to `goodKongState`).

I'd suggest just running this in a regular loop. I think it would make sense as what we want to do is to pick the first good config we find anyway.

Reply:
I wasn't aware of its multi-threading capabilities. I've replaced it with a simple loop.